### ISSUE TITLE
Registering multiple plugins means ambiguous options

### ISSUE THREAD

**Issue by jsumners** [fastify/fastify#456]
This is alluded to in the discussion on #304. Consider the following:

```js
'use strict'

const plugin1 = function (instance, opts, next) {
  console.log(opts)
  next()
}
plugin1[Symbol.for('skip-override')] = true

const plugin2 = function (instance, opts, next) {
  console.log(opts)
  next()
}
plugin2[Symbol.for('skip-override')] = true

const fastify = require('fastify')()

fastify.register([plugin1, plugin2], {foo: 'foo', bar: 'bar'}, (err) => {
  if (err) throw err
})

fastify.listen(8000, (err) => {
  if (err) throw err
  fastify.close()
})
```

The output of running that will be:

```
{ foo: 'foo', bar: 'bar' }
{ foo: 'foo', bar: 'bar' }
```

Let's assume `plugin1` and `plugin2` solve completely different problems. Let's also assume that we are writing a plugin that depends on `plugin1` and `plugin2` being registered, ad that our new plugin provides configuration options to pass configuration to each of the two required plugins. Well, unless we want to start writing a pyramid of doom, we can't do that because `plugin1` and `plugin2` share common configuration keys. Which means we are unable to effectively configure each plugin.

I am proposing the following solution:

```js
function plugin (fastify, options, next) {
  next()
}

module.exports = fp(plugin, '>= 0.34.0')
module.exports.fastifyMeta = {
  name: 'pluginName',
  version: '1.0.0'
}
```

Which will allow:

```js
const options = {
  prefix: '/foo',
  plugin1: { foo: 'foo' },
  plugin2: { foo: 'foobar' }
}

fastify.register([plugin1, plugin2], options, (err) => {
  if (err) throw err
})
```

So the `register` method will look for a key in the passed in `options` object that matches `plugin.fastifyMeta.name` and pass that key's value in as the options when registering that plugin. Otherwise, it will pass in the whole options object as it does now.

I'm not a big fan of hanging properties off the exported function. But I think the ecosystem is already too large to change the way `register` works any more than this proposal. What are your thoughts @fastify/fastify ?

**Comment by delvedor** [comment]
Probably the best solution in this case is to register the plugins separately.
```js
fastify
  .register(plugin1, opts1)
  .register(plugin2, opts2)
```
Because is not that easy understand which key are common options (such as `prefix`) or which are plugins name (then we'll have the problem of private names).
I think that we should support one and clear way to to this, even if that means that you must write few characters more.

Regarding the plugin name for stuff like the child logger, use the function name or `module.exports.meta` could be a good option.
Another option could be update `fastify-plugin` to accept a `name` parameter, and handle this internally.

**Comment by jsumners** [comment]
Okay, so if we modify my original example to:

```js
const plugin1 = function (instance, opts, next) {
  console.log(opts)
  next(Error('plugin1 error'))
}
plugin1[Symbol.for('skip-override')] = true

// snip

fastify
  .register(plugin1, {foo: 'foo'})
  .register(plugin2, {bar: 'bar'}, (err) => {
    if (err) console.error('error received!: %s', err.message)
  })

// snip
```

The `console.error` is not hit. Instead, the error gets trapped by the `if (err) throw err` within the `fastify.listen` callback. But if we revert to the original example, and add the `Error` in `plugin1`, then the error gets thrown in the `register` callback. Which is to say, the proposed solution means extra error handling; one handler per plugin.

Here's a simplified example of the proposed `register` internals:

```js
function register (plugins, options, next) {
  for (const plugin of plugins) {
    const pname = (plugin.fastifyMeta && plugin.fastifyMeta.name) ? plugin.fastifyMeta.name : null
    if (pname && options[pname]) {
      plugin(options[pname], next)
    } else {
      plugin(options, next)
    }
  }
}
```

**Comment by jsumners** [comment]
Another thing that `fastifyMeta` could provide is a way to declare plugin dependencies:

```js
module.exports = fp(foo, '>= 0.33.0')
module.exports.fastifyMeta = {
  name: 'foo-plugin',
  dependencies: [
    {name: 'fastify-cookie', version: '>=1.0.0'}
  ]
}
```

Thus, Fastify would be able to track the plugins that have been registered in the current scope, check that the required plugins have been registered to the current scope, and refuse to register the requested plugin if the requirements are not met.

See `dependencies` in https://hapijs.com/api#plugins for a less robust equivalent mechanism.

**Comment by mcollina** [comment]
I'm actually thinking that we should remove multiple plugin registration. The register chainable api is very easy to use, and an array can be easily iterated on top.

**Comment by delvedor** [comment]
@jsumners the callback after the register is not supported anymore. At the moment you should use an after callback.

```js
// this will not work
fastify.register(plugin, opts, cb)

// this will work
fastify.register(plugin, opts).after(cb)
```

Related: https://github.com/fastify/fastify/issues/460
___
I think we should handle all this work around plugin names, options and dependencies with `fastify-plugin`. We are already using it in every plugin, and it can provide us an easy way to handle breaking changes in the future, since it's under our control.

**Comment by jsumners** [comment]
@delvedor how will `fastify-plugin` know anything about the state of the destination scope of Fastify?

**Comment by delvedor** [comment]
@jsumners can you rephrase? Maybe with an example :D

**Comment by jsumners** [comment]
@delvedor look at https://www.npmjs.com/package/fastify-session . Notice that it requires `fastify-cookie` to be installed *and* registered prior to registering `fastify-session`. Also notice that it doesn't make `fastify-cookie` a hard dependency; it's left up to the user of the plugin to meet the requirement. This is not an uncommon thing to do for such plugins. And it's what I'm doing with my version of the same plugin, but mine requires `fastify-caching` as well.

So the issue is that the instance/scope of Fastify that a plugin is being registered to needs to meet certain requirements. Namely, it must have other plugins registered first. And I don't see how `fastify-plugin` can satisfy that requirement.

**Comment by delvedor** [comment]
In the same way we are passing the `skip-override` property, we can pass different options.
```js
const opts = {
  version: '>=0.30.0',
  dependencies: ['plugin1', 'plugin2']
}
module.exports = fp(plugin, opts)
```

If the problem is "I don't see how fastify-plugin can satisfy that requirement." the answer is easy.
It can't, because it cannot access the fastify instance.

But since `fastify-plugin` is under our control we can coordinate core and `fastify-plugin`, in the same way that we are handling the override property [here](https://github.com/fastify/fastify/blob/master/fastify.js#L264-L288).

**Comment by jsumners** [comment]
Okay.
